third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -356,9 +370,42 @@ where
                         .charge_drop_frame(non_ref_vals.iter())
                         .map_err(|e| set_err_info!(current_frame, e))?;
 
-                    if RTTCheck::should_perform_checks() {
-                        self.check_return_tys(&mut current_frame)
+                    // If the returning function has runtime checks on, the return types
+                    // will be on the caller stack.
+                    let caller_has_rt_checks = self
+                        .call_stack
+                        .0
+                        .last()
+                        .map(|f| RTTCheck::should_perform_checks(&f.function.function))
+                        .unwrap_or_default();
+                    let callee_has_rt_checks =
+                        RTTCheck::should_perform_checks(&current_frame.function.function);
+                    if callee_has_rt_checks {
+                        self.check_return_tys::<RTTCheck>(&mut current_frame)
                             .map_err(|e| set_err_info!(current_frame, e))?;
+                        if !caller_has_rt_checks {
+                            // The callee has pushed return types, but they aren't used by
+                            // the caller, so need to be removed.
+                            self.operand_stack
+                                .popn_tys(current_frame.function.return_tys().len() as u16)
+                                .map_err(|e| set_err_info!(current_frame, e))?;
+                        }
+                    } else if caller_has_rt_checks {
+                        // We are not runtime checking this function, but in the caller,
+                        // so we must push the return types of the function onto the type stack,
+                        // following the runtime type checking protocol.
+                        let ty_args = current_frame.function.ty_args();
+                        for mut ret_ty in current_frame.function.return_tys().iter().cloned() {
```

Comment:
I think it is better to have 2 loops -> otherwise we can clone types twice which require substitution